tools/ports/__init__.py

```diff
@@ -79,6 +79,21 @@ def dir_is_newer(dir_a, dir_b):
   return newest_a[1] > newest_b[1]
 
 
+def maybe_copy(src, dest):
+  """Just like shutil.copyfile but will do nothing if the destination already
+  exists and has the same contents as the source.
+
+  This is useful for avoiding races when building and installing different
+  flavors of the same library, which happens especially run running the test
+  suite.  In this case the library gets build multiple times but we don't want
+  to clobber the header files each time, since there is a chance another test is
+  currently executing and reading the header we are installing.
+  """
+  if os.path.exists(dest) and utils.read_file(src) == utils.read_file(dest):
+    return
+  shutil.copyfile(src, dest)
```

Comment:
Does this help on POSIX? I'm actually not sure if it's an atomic operation or not. But if it isn't, then the check above it isn't guaranteed to work either?

(And I have no idea about windows...)

Reply:
None of this is atomic, but it all happens with the cache write lock held.

The issue is that there are other processes running that don't hold the lock that are simply consuming/reading this file.  (e.g. normal compiles such as `emcc -c -sUSE_ZLIB`).    From the POV of those other processes we don't want to be removing and then overwriting these files, even though we do hold the write lock.

Reply:
I updated the comment to make it more clear exactly why we do this.

Reply:
Ah, thanks, I didn't realize this was all when the lock is held.